common/web/gesture-recognizer/src/engine/eventSequentializationQueue.ts

```diff
@@ -0,0 +1,51 @@
+import { timedPromise } from "@keymanapp/web-utils";
+
+export class EventSequentializationQueue {
+  private queue: (() => Promise<void> | void)[];
+  private defermentPromise: Promise<void>;
+
+  constructor() {
+    this.queue = [];
+  }
+
+  private setDeferment(promise: Promise<any>) {
+    this.defermentPromise = promise;
+    promise.then(() => {
+      this.defermentPromise = null;
+      this.triggerEvent();
+    });
+  }
+
+  private async triggerEvent() {
+    while(this.queue.length > 0) {
+      const functor = this.queue.shift();
+
+      // Things break _badly_ if we don't keep the queue running if errors are triggered by the functor.
+      // It's best to ignore the error and let things play out.
+      try {
+        // Is either undefined or is a Promise.
+        const result = functor();
+        // We either wait on a manual lock (from within an InputEventEngine) or a macrotask queue wait,
+        // allowing gesture-matching microtask queue Promises to complete before proceeding.
+        this.setDeferment(result ? result : timedPromise(0));
```

Comment:
If there is more than one item in the queue, then calling `setDeferment` the second time will cause the first promise to not be waited on rather than chaining. Is that intended?

Reply:
> This is the same no?

Sadly, no.  I tried to do that first, but TS fell over on the typing.  The nature of the error it gave me suggests that it doesn't recognize a function with `void` return type as returning `undefined`, which would be necessary for the direct optional chaining to work.

![image](https://github.com/keymanapp/keyman/assets/25213402/2c139f53-8b86-4947-bcef-8ad26d0d47bc)

Alternatively, it _is_ happy if I change line 27 to this:

```typescript
const result = functor() as undefined | Promise<any>;
```

Reply:
> If there is more than one item in the queue, then calling `setDeferment` the second time will cause the first promise to not be waited on rather than chaining. Is that intended?

Ah, good catch.  Looks like I messed up with my simplification from its original version in the branch that had tons of logging; will fix.  ... and that was missing proper handling in one of its branches anyway.

Reply:
> Sadly, no. I tried to do that first, but TS fell over on the typing

Such is life. I guess `void` is technically not the same as `undefined`! Original is fine then, this was just a nit.